### PR TITLE
Azure App service Manage & Deploy Task Spec changes

### PR DESCRIPTION
**Pull request by vincentdass**

### CODE REVIEW

File: Tasks/AzureAppServiceManage/azureappservicemanage.ts [hunk 1]
```diff
@@ -28,6 +28,20 @@ async function updateKuduDeploymentLog(endPoint, webAppName, resourceGroupName,
...
 }
 
+async function waitForAppServiceToStart(endPoint, resourceGroupName, webAppName, specifySlotFlag, slotName) {
+    var appServiceDetails = await azureRmUtil.getAppServiceDetails(endPoint, resourceGroupName, webAppName, specifySlotFlag, slotName);
+    if(appServiceDetails.hasOwnProperty("properties") && appServiceDetails.properties.hasOwnProperty("state"))
+    {
+        tl.debug('App Service State : ' + appServiceDetails.properties.state);
+        while(!(appServiceDetails.properties.state == "Running" || appServiceDetails.properties.state == "running"))
+        {
+            appServiceDetails = await azureRmUtil.getAppServiceDetails(endPoint, resourceGroupName, webAppName, specifySlotFlag, slotName);
```
Comment (dikhakha) [inline Tasks/AzureAppServiceManage/azureappservicemanage.ts]: do you want to add a wait() here and retry after certain intervals?

File: Tasks/AzureRmWebAppDeployment/postDeploymentScript/mainCmdFile.cmd [hunk 2]
```diff
@@ -1,8 +1,13 @@
...
 
 :: Send stdout and stderr to seperate file
-call "kuduPostDeploymentScript_%1.cmd" > "stdout_%1.txt" 2> "stderr_%1.txt"
-
 :: write return code of previous command to script_result file
 :: only return code (Numeric code) should present inside script_result
-echo %errorlevel% > "script_result_%1.txt"
\ No newline at end of file
+
+cmd /c "kuduPostDeploymentScript_%1.cmd" > "stdout_%1.txt" 2> "stderr_%1.txt" && (
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/postDeploymentScript/mainCmdFile.cmd]: Try the changes that we discussed.

File: Tasks/Common/webdeployment-common/generatewebconfig.ts [hunk 3]
```diff
@@ -17,3 +17,53 @@ function replaceMultiple(text: string, substitutions: any): string {
     }
     return text;
 }
+
```
Comment (kmkumaran) [inline Tasks/Common/webdeployment-common/generatewebconfig.ts]: Also reconcile between generateWebConfigFile and addWebConfigFile

File: Tasks/AzureAppServiceManage/azureappservicemanage.ts [hunk 4]
```diff
@@ -62,14 +76,7 @@ async function run() {
...
-                {
-                    while(!(appServiceDetails.properties.state == "Running" || appServiceDetails.properties.state == "running"))
-                    {
-                        appServiceDetails = await azureRmUtil.getAppServiceDetails(endPoint, resourceGroupName, webAppName, specifySlotFlag, slotName);
-                    }
-                }
+                await waitForAppServiceToStart(endPoint, resourceGroupName, webAppName, specifySlotFlag, slotName);
                 break;
             }
             case "Stop Azure App Service": {
```
Comment (vincentdass) [inline Tasks/AzureAppServiceManage/azureappservicemanage.ts]: Stop actually stops the App Service to refuse any request. We should be good here.

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 5]
```diff
@@ -475,6 +475,7 @@
         "ScriptStatusTimeout": "Unable to fetch script status due to timeout.",
         "PollingForFileTimeOut": "Unable to fetch script status due to timeout. You can increase the timeout limit by setting 'appservicedeploy.retrytimeout' variable to number of minutes required.",
         "InvalidPollOption": "Invalid polling option provided: %s.",
-        "MissingWebConfigParameters": "Some web.config parameters are missing"
+        "MissingWebConfigParameters": "Some web.config parameters are missing",
```
Comment (vincentdass) [inline Tasks/AzureRmWebAppDeployment/task.json]: Help Text need to be updated at many places. Will update once I get the message reviewed by @azooinmyluggage

File: Tasks/AzureRmWebAppDeployment/Strings/resources.resjson/en-US/resources.resjson [hunk 6]
```diff
@@ -140,7 +140,7 @@
   "loc.messages.ScriptStatusTimeout": "Unable to fetch script status due to timeout.",
   "loc.messages.PollingForFileTimeOut": "Unable to fetch script status due to timeout. You can increase the timeout limit by setting 'appservicedeploy.retrytimeout' variable to number of minutes required.",
   "loc.messages.InvalidPollOption": "Invalid polling option provided: %s.",
-  "loc.messages.MissingWebConfigParameters": "Some web.config parameters are missing",
+  "loc.messages.MissingAppTypeWebConfigParameters": "-appType attribute is missing in Web.config parameters.",
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/Strings/resources.resjson/en-US/resources.resjson]: Is apptype the only config param that can be missing? Should we also report if other required parameters are not present?

Comment (vincentdass) [inline Tasks/AzureRmWebAppDeployment/Strings/resources.resjson/en-US/resources.resjson]: We can override with default values.

File: Tasks/AzureAppServiceManage/azureappservicemanage.ts (hunk 1 above)
Comment (vincentdass) [inline Tasks/AzureAppServiceManage/azureappservicemanage.ts]: wait is actually not required as we just poll for the state of the app and not on any operation.

Comment (Ajay-MS) [inline Tasks/AzureAppServiceManage/azureappservicemanage.ts]: We should  check for hasProperty("properties") and hasProperty("state") after getting appServiceDetails in while loop

File: Tasks/AzureAppServiceManage/Strings/resources.resjson/en-US/resources.resjson [hunk 7]
```diff
@@ -23,6 +23,8 @@
   "loc.input.label.Slot": "Slot",
   "loc.input.label.ExtensionsList": "Install Extensions",
   "loc.input.help.ExtensionsList": "Site Extensions run on Microsoft Azure App Service. You can install set of tools as site extension and better manage your Azure App Service. The  App Service will be restarted to make sure latest changes take effect.",
+  "loc.input.label.OutputVariable": "Output variable",
+  "loc.input.help.OutputVariable": "Provide comma separated list of variables that saves the local path of the extension which can be used in subsequent tasks.",
```
Comment (RoopeshNair) [inline Tasks/AzureAppServiceManage/Strings/resources.resjson/en-US/resources.resjson]: "Provide the variable name for the local installation path for the selected extension. <br/>

Note: In case of multiple extensions selected for installation, provide comma separated list of variables that saves the local path for each of the selected extension in the order it appears in the Install Extension field"

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 8]
```diff
@@ -475,6 +475,10 @@
         "ScriptStatusTimeout": "Unable to fetch script status due to timeout.",
         "PollingForFileTimeOut": "Unable to fetch script status due to timeout. You can increase the timeout limit by setting 'appservicedeploy.retrytimeout' variable to number of minutes required.",
         "InvalidPollOption": "Invalid polling option provided: %s.",
-        "MissingWebConfigParameters": "Some web.config parameters are missing"
+        "MissingAppTypeWebConfigParameters": "-appType attribute is missing in Web.config parameters.",
```
Comment (RoopeshNair) [inline Tasks/AzureRmWebAppDeployment/task.json]: Attribute '-appType' is missing in the Web.config parameters. Valid values for '-appType' are: 'python_Bottle', 'python_Django', 'python_Flask' and 'node'. <br/>
For example, '-appType python_Bottle' (sans-quotes) in case of Python Bottle framework.

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 9]
```diff
@@ -475,6 +475,10 @@
...
+        "AutoDetectDjangoSettingsFailed": "Unable to detect DJANGO_SETTINGS_MODULE (settings.py). Please ensure that the file exists or provide the correct path in Web.config parameters input in the following format '<folder_name>.settings'",
```
Comment (RoopeshNair) [inline Tasks/AzureRmWebAppDeployment/task.json]: "Unable to detect DJANGO_SETTINGS_MODULE 'settings.py' file path. Please ensure that the 'settings.py' file exists or provide the correct path in Web.config parameter input in the following format '-DJANGO_SETTINGS_MODULE <folder_name>.settings'",

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 10]
```diff
@@ -475,6 +475,10 @@
...
+        "FailedToApplyTransformation": "Unable to apply transformation for the given package. Please follow below steps to debug the issue. \n1. Transformation is applied for MSBuild generated packge during build time. Remove <DependentUpon> Tag for each config file. \n2. Ensure that the config file and transformation file are present in the same folder inside the package.",
```
Comment (RoopeshNair) [inline Tasks/AzureRmWebAppDeployment/task.json]: "Unable to apply transformation for the given package. Verify the following. \n1. Whether the Transformation is already applied for the MSBuild generated package during build. If yes, remove the <DependentUpon> tag for each config in the csproj file and rebuild. \n2. Ensure that the config file and transformation files are present in the same folder inside the package.",

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 11]
```diff
@@ -475,6 +475,10 @@
...
+        "AutoParameterizationMessage": "ConnectionString attributes in Web.config is parameterized by default. Please be aware that transformation has no effect on connectioString attributes as the same value is overridden during deployment. You can disable the auto parameterization by setting /p:AutoParameterizationWebConfigConnectionStrings=False during MSBuild package generation.",
```
Comment (RoopeshNair) [inline Tasks/AzureRmWebAppDeployment/task.json]: ConnectionString attributes in Web.config is parameterized by default. Note that the transformation has no effect on connectionString attributes as the value is overridden during deployment by 'Parameters.xml or SetParameters" files. You can disable the auto-parameterization by setting /p:AutoParameterizationWebConfigConnectionStrings=False during MSBuild package generation."

File: Tasks/Common/webdeployment-common/ltxdomutility.ts [hunk 12]
```diff
@@ -82,7 +82,7 @@ export function getChildElementsByTagName(node, tagName) {
             }
             var liveChildNodes = getChildElementsByTagName(children[i], tagName);
             if(liveChildNodes && liveChildNodes.length > 0){
-            	liveNodes.concat(liveChildNodes);
+            	liveNodes = liveNodes.concat(liveChildNodes);
```
Comment (kmkumaran) [inline Tasks/Common/webdeployment-common/ltxdomutility.ts]: Why this change?

File: Tasks/Common/webdeployment-common/fileTransformationsUtility.ts [hunk 13]
```diff
@@ -7,9 +7,18 @@ var jsonSubstitutionUtility = require('webdeployment-common/jsonvariablesubstitu
...
 var xdtTransformationUtility = require('webdeployment-common/xdttransformationutility.js');
 
-export function fileTransformations(isFolderBasedDeployment: boolean, JSONFiles: any, xmlTransformation: boolean, xmlVariableSubstitution: boolean, folderPath: string) {
+export function fileTransformations(isFolderBasedDeployment: boolean, JSONFiles: any, xmlTransformation: boolean, xmlVariableSubstitution: boolean, folderPath: string, webDeployPkg: string) {
 
     if(xmlTransformation) {
+        if(!isFolderBasedDeployment && utility.isMSDeployPackage(webDeployPkg)) {
+            var debugMode = tl.getVariable('system.debug');
+            if(debugMode && debugMode.toLowerCase() == 'true') {
+                tl.warning(tl.loc('AutoParameterizationMessage'));
```
Comment (kmkumaran) [inline Tasks/Common/webdeployment-common/fileTransformationsUtility.ts]: Where is this message from? I can't see in task.json.

File: Tasks/Common/webdeployment-common/fileTransformationsUtility.ts [hunk 14]
```diff
@@ -7,9 +7,18 @@ var jsonSubstitutionUtility = require('webdeployment-common/jsonvariablesubstitu
 var xmlSubstitutionUtility = require('webdeployment-common/xmlvariablesubstitutionutility.js');
 var xdtTransformationUtility = require('webdeployment-common/xdttransformationutility.js');
 
-export function fileTransformations(isFolderBasedDeployment: boolean, JSONFiles: any, xmlTransformation: boolean, xmlVariableSubstitution: boolean, folderPath: string) {
+export function fileTransformations(isFolderBasedDeployment: boolean, JSONFiles: any, xmlTransformation: boolean, xmlVariableSubstitution: boolean, folderPath: string, webDeployPkg: string) {
```
Comment (kmkumaran) [inline Tasks/Common/webdeployment-common/fileTransformationsUtility.ts]: why do you need webDeployPkg itself instead of passing a Boolean of isMSDeployPackage?

File: Tasks/Common/webdeployment-common/fileTransformationsUtility.ts (hunk 13 above)
Comment (vincentdass) [inline Tasks/Common/webdeployment-common/fileTransformationsUtility.ts]: Message is available in [AzureRmWebAppDeployment/task.json](https://github.com/Microsoft/vsts-tasks/blob/73df31463d98bdcde73c5441edeed76f83371f8a/Tasks/AzureRmWebAppDeployment/task.json#L481)

File: Tasks/AzureAppServiceManage/extensionmanage.ts [hunk 15]
```diff
@@ -64,19 +65,48 @@ export async function installExtension(publishingProfile, extension: string) {
...
 }
+
+function getExtensionLocalPath(extensionInfo: JSON): string {
+    var extensionId: string = extensionInfo['id'];
+    var homeDir = "D:\\home\\";
+
+    if(extensionId.startsWith('python2')) {
+        return homeDir + "Python27";
+    }
+    else if(extensionId.startsWith('python351') || extensionId.startsWith('python352')) {
```
Comment (dikhakha) [inline Tasks/AzureAppServiceManage/extensionmanage.ts]: this may not be required as we are not showing option to install 3.5 < python version  < 3.5.3 ?

Comment (vincentdass) [inline Tasks/AzureAppServiceManage/extensionmanage.ts]: Additional check, if we add them.
